### PR TITLE
AXI DMA Driver bugfixes/improvements

### PR DESCRIPTION
**Pull request by jeremytrimble**
I've sent the following to git@xilinx.com.  Posting here on github as well for the sake of open discussion.  It would be nice if in the future Xilinx could set up a subscribable mailing list that would allow patches to be discussed in the open.

---

Hello,

This patch series addresses a number of issues I discovered while using the
latest xilinx_dma.c driver in my Zynq project.  Changes were tested with both
S2MM and MM2S AXI DMA cores.

A number of issues are addressed, but the two most significant are:
1. Removal of incorrect interrupt coalescing code.
   - See also the discussion on the Xilinx forums at [1].
2. Re-introduction of a dma_pool for segments.
   - "Segments" are also referred to as "Buffer Descriptors in Xilinx PG021.
   - My application requires more than 255 descriptors at any given time.
   - The approach used here eliminates the 255 limitation without requiring
     the DMA to be halted and re-started when new segments are issued to the
     hardware.

I also found that under some circumstances, even with IrqThreshold set to 1, it
is possible to "miss" the interrupt for the last-issued segment (the segment
which is pointed to by the TAILDESC register), causing the driver to "hang" in
a state where all outstanding buffer descriptors have been processed by the
hardware (and thus no additional interrupts are coming), but software doesn't
know that, and thus will never issue future segments to the hardware.

The only way I saw to get around this issue without compromising throughput was
to implement periodic re-polling from within software using a kernel timer.

[1] https://forums.xilinx.com/t5/Embedded-Processor-System-Design/AXI-DMA-IrqThreshold-Clarification/td-p/671737

Regards,
Jeremy Trimble

### CODE REVIEW

**Comment by ismell** [comment]
How about updating axidmatest to test some of these edge cases?
* Add a test that has a larger receive buffer then the tx buffer so you can verify the interrupt coalescing.
* Add a test that allocates more then 255 BDs and pushes them through.

Missing an interrupt is concerning. I'm wondering if the interrupt handler is acking too early or something.

Also to the xilinx folks, How about using the PRs to discuss and then just manually merging in the PR into mainline?
https://github.com/spring-projects/spring-framework/wiki/Manually-merging-pull-requests has a nice workflow for manual merging.

**Comment by Lufft-GitHub** [comment]
Thanks for these fixes.
Did you also get cyclic mode working correctly? We are still facing some problems.

BTW: It seems that one descriptor is not getting freed correctly on driver exit. I always receive an exception when unloading the xilinx_dma driver. Unfortunatly I dind't investigate in this in more detail yet.

**Comment by Lufft-GitHub** [comment]
In my case cyclic mode didn't work as I didn't enable "Micro mode" in my Axi DMA. In this case Axi DMA will *only* use the TAILDESC pointer to detect end of descriptor chain. As the manual says, in cycling mode this should only be a dummy value to initiate transfer.

**Comment by jeremytrimble** [comment]
I wasn't able to test these changes under cyclic mode.  I'm currently off work for a bit for some personal business, but I'll take a look into it when I return.  Were you able to get your application working using cyclic mode with "micro DMA" enabled?

**Comment by Lufft-GitHub** [comment]
My application is working without micro DMA. I simply write a dummy value into TAILDESC register to start dma transfer. Thanks!

**Comment by jeras** [comment]
We got cyclic DMA to partially work, which means the HW is running in cyclic mode, but SW is not handling IRQ and exit cleanup correctly.

There are two bugs in HW config:
1. the last descriptor should be linked to the first one but it is not
 Our solution is incomplete, it does not check if a segment is larger then the HW counter size and therefore divided into smaller descriptors.
2. TAILDESC should be loaded with a dummy value which must not be the address of any of the descriptors
 We are using the address of the last descriptor + 0x80, but if the memory for descriptors is not continuous, this might still point to a descriptor, I am not sure what a safe dummy value would be.

Then there are issues with the SW interrupt handler. The interrupt routine should do nothing else then handle the callback, instead it does a convoluted attempt to free descriptors, which should be kept, since they are still being used. We skipped this code for cyclic mode, then there is code in the tasklet which again performs some cleanup, and later attempts to run a callback on a completed descriptor, but again it should be running the callback for an active descriptor. We improvised there too.

We have two open issues.
1. Our changes are skipping most of the descriptor cleanup code, so we can not do more than one run, since without a proper cleanup the driver could not be rerun properly. I think the cleanup code should be placed inside terminate_all, since I assume this is the only way to end a cyclic transfer (if you disagree I would like to discuss it).
2. It seems we are not getting IOC interrupts without a packet end (TLAST on AXI stream), which is not what I would expect from reading the documentation and clarifications on various forums. How is streaming supposed to work without an IRQ after every segment? I might be wrong, could anybody suggest a proper way for counting interrupts (it might be common knowledge, but I am a HW developer, this is my first kernel hacking experience).

So our current state is, we can see the DMA is running in cycles (by monitoring registers and descriptors), but we can run a single test on each reboot, and we are not sure when to expect interrupts.

**Comment by jeras** [comment]
our current patch:

```patch
diff --git a/drivers/dma/xilinx/xilinx_dma.c b/drivers/dma/xilinx/xilinx_dma.c
index 9902daf..e68e213 100644
--- a/drivers/dma/xilinx/xilinx_dma.c
+++ b/drivers/dma/xilinx/xilinx_dma.c
@@ -567,7 +567,8 @@ static void xilinx_dma_halt(struct xilinx_dma_chan *chan)
 	u32 val;
 
 	chan->ctrl_reg &= ~XILINX_DMA_CR_RUNSTOP_MASK;
-	dma_ctrl_write(chan, XILINX_DMA_REG_CONTROL, chan->ctrl_reg);
+	dma_ctrl_write(chan, XILINX_DMA_REG_CONTROL, chan->ctrl_reg |
+		       XILINX_DMA_CR_RESET_MASK);
 
 	/* Wait for the hardware to halt */
 	err = xilinx_dma_poll_timeout(chan, XILINX_DMA_REG_STATUS, val,
@@ -653,13 +654,23 @@ static void xilinx_dma_start_transfer(struct xilinx_dma_chan *chan)
 
 	/* Start the transfer */
 	if (chan->has_sg) {
+		if (chan->cyclic) {
+#ifdef CONFIG_PHYS_ADDR_T_64BIT
+			dma_ctrl_writeq(chan, XILINX_DMA_REG_TAILDESC,
+				       tail_segment->phys + 0x80);
+#else
+			dma_ctrl_write(chan, XILINX_DMA_REG_TAILDESC,
+				       tail_segment->phys + 0x80);
+#endif
+		} else {
 #ifdef CONFIG_PHYS_ADDR_T_64BIT
-		dma_ctrl_writeq(chan, XILINX_DMA_REG_TAILDESC,
-			       tail_segment->phys);
+			dma_ctrl_writeq(chan, XILINX_DMA_REG_TAILDESC,
+				       tail_segment->phys);
 #else
-		dma_ctrl_write(chan, XILINX_DMA_REG_TAILDESC,
-			       tail_segment->phys);
+			dma_ctrl_write(chan, XILINX_DMA_REG_TAILDESC,
+				       tail_segment->phys);
 #endif
+		}
 	} else {
 		struct xilinx_dma_tx_segment *segment;
 		struct xilinx_dma_desc_hw *hw;
@@ -786,7 +797,7 @@ static irqreturn_t xilinx_dma_irq_handler(int irq, void *data)
 	if (status & XILINX_DMA_XR_IRQ_DELAY_MASK)
 		dev_dbg(chan->dev, "Inter-packet latency too long\n");
 
-	if (status & XILINX_DMA_XR_IRQ_IOC_MASK) {
+	if ((status & XILINX_DMA_XR_IRQ_IOC_MASK) && !(chan->cyclic)) {
 		spin_lock(&chan->lock);
 		xilinx_dma_complete_descriptor(chan);
 		chan->idle = true;
@@ -804,9 +815,25 @@ static irqreturn_t xilinx_dma_irq_handler(int irq, void *data)
  */
 static void xilinx_dma_do_tasklet(unsigned long data)
 {
+	struct xilinx_dma_tx_descriptor *desc;
+	dma_async_tx_callback callback;
+	void *callback_param;
+
 	struct xilinx_dma_chan *chan = (struct xilinx_dma_chan *)data;
 
-	xilinx_dma_chan_desc_cleanup(chan);
+	if (chan->cyclic) {
+		desc = list_first_entry(&chan->active_list,
+			struct xilinx_dma_tx_descriptor, node);
+
+		callback = desc->async_tx.callback;
+		callback_param = desc->async_tx.callback_param;
+		if (callback) {
+			callback(callback_param);
+			printk("tasklet\n");
+		}
+	} else {
+		xilinx_dma_chan_desc_cleanup(chan);
+	}
 }
 
 /**
@@ -1013,6 +1040,7 @@ static struct dma_async_tx_descriptor *xilinx_dma_prep_dma_cyclic(
 	struct xilinx_dma_chan *chan = to_xilinx_chan(dchan);
 	struct xilinx_dma_tx_descriptor *desc;
 	struct xilinx_dma_tx_segment *segment;
+	struct xilinx_dma_tx_segment *head_desc, *tail_desc;
 	size_t copy, sg_used;
 	unsigned int num_periods;
 	int i;
@@ -1086,6 +1114,10 @@ static struct dma_async_tx_descriptor *xilinx_dma_prep_dma_cyclic(
 		segment->hw.control |= XILINX_DMA_BD_EOP;
 	}
 
+	head_desc = list_first_entry(&desc->segments, struct xilinx_dma_tx_segment, node);
+	tail_desc = list_last_entry (&desc->segments, struct xilinx_dma_tx_segment, node);
+	tail_desc->hw.next_desc = (u32) head_desc->phys;
+
 	return &desc->async_tx;
 
 error:
```

**Comment by jeras** [comment]
and this is the userspace app we use to debug desriptors, it is based on `devmem2`.

```c
/*
 * devmem2.c: Simple program to read/write from/to any location in memory.
 *
 *  Copyright (C) 2000, Jan-Derk Bakker (jdb@lartmaker.nl)
 *
 *
 * This software has been developed for the LART computing board
 * (http://www.lart.tudelft.nl/). The development has been sponsored by
 * the Mobile MultiMedia Communications (http://www.mmc.tudelft.nl/)
 * and Ubiquitous Communications (http://www.ubicom.tudelft.nl/)
 * projects.
 *
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <errno.h>
#include <signal.h>
#include <fcntl.h>
#include <ctype.h>
#include <termios.h>
#include <sys/types.h>
#include <sys/mman.h>
#include <stdint.h>
  
#define FATAL do { fprintf(stderr, "Error at line %d, file %s (%d) [%s]\n", \
  __LINE__, __FILE__, errno, strerror(errno)); exit(1); } while(0)
 
#define MAP_SIZE 4096UL
#define MAP_MASK (MAP_SIZE - 1)

int main(int argc, char **argv) {
    int fd;
    void *map_base, *virt_addr; 
	off_t target;
	
	if(argc < 2) {
		fprintf(stderr, "\nUsage:\t%s { address } [ type [ data ] ]\n"
			"\taddress : memory address to act upon\n"
			"\ttype    : access operation type : [b]yte, [h]alfword, [w]ord\n"
			"\tdata    : data to be written\n\n",
			argv[0]);
		exit(1);
	}
	target = strtoul(argv[1], 0, 0);

    if((fd = open("/dev/mem", O_RDWR | O_SYNC)) == -1) FATAL;
    printf("/dev/mem opened.\n"); 
    fflush(stdout);
    
    /* Map one page */
    map_base = mmap(0, MAP_SIZE, PROT_READ | PROT_WRITE, MAP_SHARED, fd, target & ~MAP_MASK);
    if(map_base == (void *) -1) FATAL;
    printf("Memory mapped at address %p.\n", map_base); 
    fflush(stdout);
    

    struct xilinx_dma_desc_hw {
        uint32_t next_desc;
        uint32_t next_desc_msb;
        uint32_t buf_addr;
        uint32_t buf_addr_msb;
        uint32_t pad1;
        uint32_t pad2;
        uint32_t control;
        uint32_t status;
        uint32_t app[5];
    };
    for (int i=0; i<4; i++) {
            virt_addr = map_base + (target & MAP_MASK) + i * 0x80;

	    printf ("Descriptor %d at address 0x%08X (%p):\n", i, (int) target, virt_addr); 
            struct xilinx_dma_desc_hw *hw;
            hw = virt_addr;
            printf ("descriptor @:%p\n", hw);
            printf ("descriptor next_desc: %08x %08x\n", hw->next_desc_msb, hw->next_desc);
            printf ("descriptor buf_addr : %08x %08x\n", hw->buf_addr_msb , hw->buf_addr );
            printf ("descriptor pad 1, 2 : %08x %08x\n", hw->pad1         , hw->pad2     );
            printf ("descriptor ctl, sts : %08x %08x\n", hw->control      , hw->status   );
            printf ("descriptor app      : %08x %08x %08x %08x %08x\n", hw->app[0], hw->app[1], hw->app[2], hw->app[3], hw->app[4]);
    }
    fflush(stdout);
    close(fd);
    return 0;
}
```

**Comment by Lufft-GitHub** [comment]
> 1. Our changes are skipping most of the descriptor cleanup code, so we can not do more than one run, since without a proper cleanup the driver could not be rerun properly. I think the cleanup code should be placed inside terminate_all, since I assume this is the only way to end a cyclic transfer (if you disagree I would like to discuss it).

We have noticed the same proble. This comes from remaining descriptors in done and pending list. As it is required to reinit active list again we simply call xilinx_dma_free_descriptors() within terminate_all()

> 2. It seems we are not getting IOC interrupts without a packet end (TLAST on AXI stream), which is not what I would expect from reading the documentation and clarifications on various forums. How is streaming supposed to work without an IRQ after every segment? I might be wrong, could anybody suggest a proper way for counting interrupts (it might be common knowledge, but I am a HW developer, this is my first kernel hacking experience).
In nearly every case the IOC interrupt is not triggered correctly, the reson from that comes from an inproper tlast generation! To realy get a stable and working Axi-DMA application, best advice is to keep the size of each dma packet (from start byte to tlast generation) synchroniously to the size of each descriptor/cycle!

Our application meanwhile is able to push data from FPGA to an streaming server (via ethernet) without any package loss (@ 1MBit/s). To make sure that we don't miss any package we have added a sequence counter in front of every DMA cycle!

**Comment by jeras** [comment]
Could you clarify regarding IOC interrupts not being triggered correctly? You mentioned improper tlast generation, is this a MM2S bug? I forgot to tell all my comments are regarding RX DMA, we have not tested TX much yet. Is there a AXI DMA errata (for the HDL) or at least a forum post describing this issues?

We can provide a `TLAST` signal at the end of each segment for S2MM, but we intended to use `TLAST` to indicate the end of a transfer, and ignore all other segment IOC interrupts. Is at least the descriptor status register `EOF` bit reliable?

Regarding the sequence counter you mentioned, do you only use to check if the DMA transfer length is correct, or you also generate `TLAST` at the end of segments with it?

**Comment by andreamerello** [comment]
Hello,

Could you please explain what's the purpose of commit 922fc54115860b12f8d88fcaf282c615ade0185c
"xilinx_dma: Handle terminate_all gracefully." ?

It looks like it would end up in causing the completion callback to be called also for uncompleted transfers; quoting DOC Documentation/dmaengine/client.txt

    int dmaengine_terminate_all(struct dma_chan *chan)

    This causes all activity for the DMA channel to be stopped, and may
    discard data in the DMA FIFO which hasn't been fully transferred.
    No callback functions will be called for any incomplete transfers.

**Comment by jeremytrimble** [comment]
@andreamerello -- You're right, I guess it shouldn't invoke the callbacks (though I find the API hard to use in that case).  I'll remove that part when I update this pull request (may be a while).

**Comment by andreamerello** [comment]
Also, I don't understand what's the point of 5d7d728ba9924846e46d88897a076b6a5bd1e069 (Always check for completed descriptors).

I would assume that we actually get the IOC flag every time the HW completes a descriptor. Is that a workaround for some IP bug? (note that in case of other interrupts, like errors, it seems that xilinx_dma_complete_descriptor() would not do anything on descriptors that are not completed, so it seems pointless to me calling it on other events.)

About the conversion spin_lock -> spin_lock_irqsave, I *think* it is not needed, and that it just introduces some extra overhead, because AFAIK it is only useful in order to avoid deadlock when other code portions that may hold the lock are interrupted by the ISR (so we disable HW irqs to avoid this), but as long as you are in the ISR, no any other portion of the driver should interrupt the ISR.

Am I missing something ? :)

**Comment by andreamerello** [comment]
BTW I've some fixes on my repo [](https://github.com/andreamerello/linux-zynq-stable/tree/z2016-04-xilinx-dma-fixes )

With this I successfully use DMA in SG mode. I've not tested cyclic mode yet. 

I'll probably issue a pull-req as soon as I'll accomplish this task (but that may take a while).

**Comment by andreamerello** [comment]
I've opened a pull-request #92 that includes also (other than various fixes by me) selected commit from this pull-request, rebased on the current xilinx_dma driver. 

Tested in SG mode only. RFT for cyclic/interleaved, please :)
